pennylane/templates/layers/quantum_number_preserving_u2.py

```diff
@@ -0,0 +1,216 @@
+            dev = qml.device('default.qubit', wires=qubits)
+
+            # Define the ansatz
+            ansatz = partial(QuantumNumberPreservingU2, init_state=ref_state, pi_gate_include=True)
+
+            # Define the cost function
+            cost_fn = qml.ExpvalCost(ansatz, h, dev)
+
+            # Compute the expectation value of 'h'
+            layers = 2
+            params = qml.init.quantum_number_preserving_u2_normal(layers, qubits)
+            print(cost_fn(params))
+
+        **Parameter shape**
+
+        The shape of the weights argument can be computed by the static method
+        :meth:`~.QuantumNumberPreservingU2.shape` and used when creating randomly
+        initialised weight tensors:
+
+        .. code-block:: python
+
+            shape = QuantumNumberPreservingU2.shape(n_layers=2, n_wires=2)
+            weights = np.random.random(size=shape)
+
+
+    """
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(
+        self, weights, wires=None, init_state=None, pi_gate_include=False, do_queue=True, id=None
+    ):
+
+        if len(wires) < 4:
+            raise ValueError(
+                "This template requires the number of qubits to be greater than four; got {}".format(
+                    len(wires)
+                )
+            )
+        if len(wires) % 2:
+            raise ValueError(
+                "This template requires the number of qubits to be multiple of 2; got {}".format(
+                    len(wires)
+                )
+            )
+
+        self.M = len(wires) // 2
+
+        self.qwires = [
+            wires[i : i + 4] for i in range(0, len(wires), 4) if len(wires[i : i + 4]) == 4
+        ]
+        if self.M > 2:
+            self.qwires += [
+                wires[i : i + 4] for i in range(2, len(wires), 4) if len(wires[i : i + 4]) == 4
+            ]
+
+        shape = qml.math.shape(weights)
+
+        if len(shape) != 3:
+            raise ValueError(f"Weights tensor must be 3-dimensional; got shape {shape}")
+
+        if shape[1] != len(self.qwires):
+            raise ValueError(
+                f"Weights tensor must have second dimension of length {len(self.qwires)}; got {shape[1]}"
+            )
+
+        if shape[2] != 2:
+            raise ValueError(
+                f"Weights tensor must have third dimension of length 2; got {shape[2]}"
+            )
+
+        self.n_layers = shape[0]
+        # we can extract the numpy representation here
+        # since init_state can never be differentiable
+        if init_state is None:
+            raise ValueError(f"Inital state should be provided; got {init_state}")
+        self.init_state = qml.math.toarray(init_state)
+
+        self.pi_gate = pi_gate_include
+
+        super().__init__(weights, wires=wires, do_queue=do_queue, id=id)
+
+    def expand(self):
+
+        with qml.tape.QuantumTape() as tape:
+
+            qml.templates.BasisEmbedding(self.init_state, wires=self.wires)
```

Comment:
You can use `BasisState` instead.

Reply:
I had tried using `BasisState` but it lead to the failure of the `test_adjoint` test case, quoting Basis state cannot be used after other gates. So, I decided to stick with `qml.templates.BasisEmbedding`.